File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/shared/content_detail/ctas/ActionRowCta.kt
@file:Suppress("PackageNaming", "MagicNumber", "LongMethod", "LongParameterList")

package app.pantopus.android.ui.screens.shared.content_detail.ctas

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing

/**
 * P6.5 — Per-kind sticky CTA for the Public Profile detail. The kind model
 * lives in [ActionRowCtaKind]:
 *
 * - `Persona` surfaces a single primary "Follow" CTA.
 * - `Local` surfaces a primary "Message" CTA + an outline "Connect" CTA.
 *
 * Sticky footer CTA. When [kind] is null the slot renders nothing,
 * preserving the legacy "actions live in the body" layout.
 */
@Composable
fun ActionRowCta(kind: ActionRowCtaKind? = null) {
    when (kind) {
        null -> Unit
        is ActionRowCtaKind.Persona ->
            PersonaFollowFooter(
                inFlight = kind.followInFlight,
                isFollowing = kind.isFollowing,
                onFollow = kind.onFollow,
            )
        is ActionRowCtaKind.Local ->
            LocalMessageConnectFooter(
                messageInFlight = kind.messageInFlight,
                connectInFlight = kind.connectInFlight,
                isConnectRequested = kind.isConnectRequested,
                onMessage = kind.onMessage,
                onConnect = kind.onConnect,
            )
    }
}

@Composable
private fun PersonaFollowFooter(
    inFlight: Boolean,
    isFollowing: Boolean,
    onFollow: () -> Unit,
) {
    androidx.compose.foundation.layout.Column(
        modifier =
            Modifier
                .fillMaxWidth()
                .background(PantopusColors.appSurface),
    ) {
        HorizontalDivider(color = PantopusColors.appBorderSubtle, thickness = 1.dp)
        Box(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .padding(Spacing.s4),
        ) {
            Row(
                modifier =
                    Modifier
                        .fillMaxWidth()
                        .testTag("publicProfileFollowCta")
                        .clip(RoundedCornerShape(Radii.lg))
                        .background(PantopusColors.primary600)
                        .alpha(if (inFlight) 0.7f else 1f)
                        .heightIn(min = 48.dp)
                        .clickable(enabled = !inFlight, onClick = onFollow)
                        .semantics { contentDescription = if (isFollowing) "Following" else "Follow" },
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.Center,
            ) {
                PantopusIconImage(
                    icon = if (isFollowing) PantopusIcon.Check else PantopusIcon.Plus,
                    contentDescription = null,
                    size = Radii.xl,
                    tint = PantopusColors.appTextInverse,
                )
                Box(modifier = Modifier.padding(start = Spacing.s2)) {
                    Text(
                        text = if (isFollowing) "Following" else "Follow",
                        fontSize = 14.sp,
                        fontWeight = FontWeight.Bold,
                        color = PantopusColors.appTextInverse,
                    )
                }
            }
        }
    }
}

@Composable
private fun LocalMessageConnectFooter(
    messageInFlight: Boolean,
    connectInFlight: Boolean,
    isConnectRequested: Boolean,
    onMessage: () -> Unit,
    onConnect: () -> Unit,
) {
    androidx.compose.foundation.layout.Column(
        modifier =
            Modifier
                .fillMaxWidth()
                .background(PantopusColors.appSurface),
    ) {
        HorizontalDivider(color = PantopusColors.appBorderSubtle, thickness = 1.dp)
        Row(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .padding(Spacing.s4),
            horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            // Connect (outline)
            Row(
                modifier =
                    Modifier
                        .weight(1f)
                        .testTag("publicProfileConnectCta")
                        .clip(RoundedCornerShape(Radii.lg))
                        .background(PantopusColors.appSurface)
                        .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.lg))
                        .alpha(if (connectInFlight) 0.7f else 1f)
                        .heightIn(min = 48.dp)
                        .clickable(enabled = !connectInFlight, onClick = onConnect)
                        .semantics {
                            contentDescription = if (isConnectRequested) "Connection requested" else "Connect"
                        },
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.Center,
            ) {
                PantopusIconImage(
                    icon = if (isConnectRequested) PantopusIcon.Check else PantopusIcon.UserPlus,
                    contentDescription = null,
                    size = Radii.xl,
                    tint = PantopusColors.appText,
                )
                Box(modifier = Modifier.padding(start = Spacing.s2)) {
                    Text(
                        text = if (isConnectRequested) "Requested" else "Connect",
                        fontSize = 14.sp,
                        fontWeight = FontWeight.SemiBold,
                        color = PantopusColors.appText,
                    )
                }
            }
            // Message (primary)
            Row(
                modifier =
                    Modifier
                        .weight(1f)
                        .testTag("publicProfileMessageCta")
                        .clip(RoundedCornerShape(Radii.lg))
                        .background(PantopusColors.primary600)
                        .alpha(if (messageInFlight) 0.7f else 1f)
                        .heightIn(min = 48.dp)
                        .clickable(enabled = !messageInFlight, onClick = onMessage)
                        .semantics { contentDescription = "Message" },
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.Center,
            ) {
                PantopusIconImage(
                    icon = PantopusIcon.MessageSquare,
                    contentDescription = null,
                    size = Radii.xl,
                    tint = PantopusColors.appTextInverse,
                )
                Box(modifier = Modifier.padding(start = Spacing.s2)) {
                    Text(
                        text = "Message",
                        fontSize = 14.sp,
                        fontWeight = FontWeight.Bold,
                        color = PantopusColors.appTextInverse,
                    )
                }
            }
        }
    }
}

File: frontend/apps/ios/Pantopus/Features/Shared/ContentDetail/CTAs/ActionRowCTA.swift
//
//  ActionRowCTA.swift
//  Pantopus
//
//  `action_row` CTA slot for the Public profile detail. P6.5
//  differentiates the sticky footer per profile kind:
//
//  - Persona profiles surface a single primary "Follow" CTA.
//  - Local profiles surface a primary "Message" CTA + an outline
//    "Connect" CTA.
//
//  The CTA tracks the matching `PublicProfileActionState` so the button
//  reflects in-flight / succeeded poses without re-fetching the profile.
//

import SwiftUI

/// Per-kind sticky CTA for the Public profile detail. The body is empty
/// when no kind is supplied so legacy callers (and the empty-state
/// path) keep rendering the existing no-footer layout.
@MainActor
public struct ActionRowCTA: View {
    public enum Kind {
        case persona(followState: PublicProfileActionState, onFollow: @MainActor () -> Void)
        case local(
            messageState: PublicProfileActionState,
            connectState: PublicProfileActionState,
            onMessage: @MainActor () -> Void,
            onConnect: @MainActor () -> Void
        )
    }

    private let kind: Kind?

    public init(kind: Kind? = nil) {
        self.kind = kind
    }

    public var body: some View {
        switch kind {
        case .none:
            EmptyView()
        case let .persona(state, onFollow):
            PersonaFollowFooter(state: state, onFollow: onFollow)
        case let .local(messageState, connectState, onMessage, onConnect):
            LocalMessageConnectFooter(
                messageState: messageState,
                connectState: connectState,
                onMessage: onMessage,
                onConnect: onConnect
            )
        }
    }
}

@MainActor
private struct PersonaFollowFooter: View {
    let state: PublicProfileActionState
    let onFollow: @MainActor () -> Void

    var body: some View {
        Button(action: onFollow) {
            HStack(spacing: Spacing.s2) {
                Icon(state == .succeeded ? .check : .plus, size: 16, color: Theme.Color.appTextInverse)
                Text(state == .succeeded ? "Following" : "Follow")
                    .font(.system(size: PantopusTextStyle.small.size, weight: .bold))
                    .foregroundStyle(Theme.Color.appTextInverse)
            }
            .frame(maxWidth: .infinity, minHeight: 48)
            .background(Theme.Color.primary600)
            .clipShape(RoundedRectangle(cornerRadius: Radii.lg, style: .continuous))
        }
        .buttonStyle(.plain)
        .disabled(state == .inFlight)
        .opacity(state == .inFlight ? 0.7 : 1)
        .padding(Spacing.s4)
        .frame(maxWidth: .infinity)
        .background(Theme.Color.appSurface)
        .overlay(alignment: .top) {
            Rectangle().fill(Theme.Color.appBorderSubtle).frame(height: 1)
        }
        .accessibilityIdentifier("publicProfileFollowCta")
        .accessibilityLabel(state == .succeeded ? "Following" : "Follow")
    }
}

@MainActor
private struct LocalMessageConnectFooter: View {
    let messageState: PublicProfileActionState
    let connectState: PublicProfileActionState
    let onMessage: @MainActor () -> Void
    let onConnect: @MainActor () -> Void

    var body: some View {
        HStack(spacing: Spacing.s2) {
            Button(action: onConnect) {
                HStack(spacing: Spacing.s2) {
                    Icon(
                        connectState == .succeeded ? .check : .userPlus,
                        size: 16,
                        color: Theme.Color.appText
                    )
                    Text(connectState == .succeeded ? "Requested" : "Connect")
                        .font(.system(size: PantopusTextStyle.small.size, weight: .semibold))
                        .foregroundStyle(Theme.Color.appText)
                }
                .frame(maxWidth: .infinity, minHeight: 48)
                .background(Theme.Color.appSurface)
                .overlay(
                    RoundedRectangle(cornerRadius: Radii.lg, style: .continuous)
                        .stroke(Theme.Color.appBorder, lineWidth: 1)
                )
                .clipShape(RoundedRectangle(cornerRadius: Radii.lg, style: .continuous))
            }
            .buttonStyle(.plain)
            .disabled(connectState == .inFlight)
            .opacity(connectState == .inFlight ? 0.7 : 1)
            .accessibilityIdentifier("publicProfileConnectCta")
            .accessibilityLabel(connectState == .succeeded ? "Connection requested" : "Connect")

            Button(action: onMessage) {
                HStack(spacing: Spacing.s2) {
                    Icon(.messageSquare, size: 16, color: Theme.Color.appTextInverse)
                    Text("Message")
                        .font(.system(size: PantopusTextStyle.small.size, weight: .bold))
                        .foregroundStyle(Theme.Color.appTextInverse)
                }
                .frame(maxWidth: .infinity, minHeight: 48)
                .background(Theme.Color.primary600)
                .clipShape(RoundedRectangle(cornerRadius: Radii.lg, style: .continuous))
            }
            .buttonStyle(.plain)
            .disabled(messageState == .inFlight)
            .opacity(messageState == .inFlight ? 0.7 : 1)
            .accessibilityIdentifier("publicProfileMessageCta")
            .accessibilityLabel("Message")
        }
        .padding(Spacing.s4)
        .frame(maxWidth: .infinity)
        .background(Theme.Color.appSurface)
        .overlay(alignment: .top) {
            Rectangle().fill(Theme.Color.appBorderSubtle).frame(height: 1)
        }
    }
}
